File: services/marketplace/src/inbox/InboxEvent.ts
// Mongo's mirror of Core's inbox_event table — "have we already processed this event",
// one document per (event, consumer).
import { Document, Schema, model } from "mongoose";

export interface InboxEventDocument extends Document {
  eventId: string;
  consumer: string;
  receivedAt: Date;
}

const inboxEventSchema = new Schema<InboxEventDocument>({
  eventId: { type: String, required: true },
  consumer: { type: String, required: true },
  receivedAt: { type: Date, required: true, default: () => new Date() },
});

inboxEventSchema.index({ eventId: 1, consumer: 1 }, { unique: true });

export default model<InboxEventDocument>("InboxEvent", inboxEventSchema);

File: services/core/src/main/java/com/dannest/inbox/InboxEvent.java
package com.dannest.inbox;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * "Have we already processed this event" — one row per (event, consumer). See
 * {@link Idempotency#tryClaim} for the actual insert-or-skip usage; this entity only
 * exists to give that a JPA-mapped table to insert into.
 */
@Entity
@Table(name = "inbox_event")
@IdClass(InboxEventId.class)
@Getter
@NoArgsConstructor
@AllArgsConstructor
public class InboxEvent {

    @Id
    @Column(name = "event_id")
    private UUID eventId;

    @Id
    @Column(name = "consumer")
    private String consumer;

    @Column(name = "received_at", nullable = false)
    private Instant receivedAt;
}
